### ISSUE TITLE
More rigorous validation for `callable-schema`

### ISSUE THREAD

**Issue by kc0506** [pydantic/pydantic#10340]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

Currently the `core_schema.callable_schema` only checks if an input is callable, so the following works:
```py
class A(BaseModel):
    f: Callable[[], int]

def f() -> str:
    return 1
    
A(f=f)  # no error
```

With the help of `inspect.signature`, we can check the number of arguments, the annotations (if input is a function), etc., so that the above will throw an error.

### Affected Components

- [X] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by Viicos** [comment]
Seems like a valid feature request. Probably requires a substantial amount of work (validating, args, kwargs, positional only, paramspecs? typevartuples?)

**Comment by kc0506** [comment]
> Probably requires a substantial amount of work (validating, args, kwargs, positional only, paramspecs? typevartuples?)

Yes, there seem to be many cases considering the flexibility of signatures in Python. 

[Variance of types](https://peps.python.org/pep-0483/#covariance-and-contravariance) may also be relevant.

**Comment by sydney-runkle** [comment]
Sounds like a great improvement!
